## Analysis of the non-allelic homologous recombination events using a k-mer counting approach

This package aims to predict non-allelic homologous recombination (NAHR) events in ORFs and other genetic elements between genomes.

**What is the input of the analysis**

We need the sequences of query genes, and the sequences of reference genes, and the table of the homolog pairs between query genes and reference genes. The ortholog of the query gene in the reference genes must have the same name with the query gene.


**How to identify NAHR events**

For each query sequence and all the reference sequences homologous to the query sequence, we performed k-mer counting. K-mers which are shared with only between the query and the paralog, but not with the corresponding ortholog in reference, indicate SNVs converted from the paralog, resulting from NAHR.


**Gerenal Model for NAHR identification**

<details>
  <summary>Click to expand!</summary>
For a given query sequence and a set of reference sequences, we speculated that the query sequence is generated by 1) minimal number of mutations; 2) minimal number of non-overlapping sequential NAHR events. By means of minimal number of mutations, all the SNVs between the query and the ortholog that can be the result of conversion from NAHR with a paralog are considered to be recombinant SNVs, rather than mutations relative to the ortholog. By means of minimal number of non-overlapping sequential NAHR events: 1) The query sequence will be separated into regions based on the source of the recombination event, each of which represents one NAHR event or the original ortholog. 2) None of the two adjacent regions are derived from the a shared reference gene, because we can decrease the number of events by combining the two regions. 3) Furthermore, the analysis also works without an identified ortholog. The "ortholog" is dynamically identified and updated during the k-mer scanning along the query sequence (details in the following sections).  


We further defined minimal recombination regions and the homologous flanking regions. Recombination regions are regions containing sequences shared only with the query and the corresponding reference sequence(s). The minimal recombinations have two levels of definition: 1) Minimal number of recombination regions. There is one minimal recombination region within one recombination region. The number of minimal recombination region is minimized because the number of recombination region is minimized. 2) Minimal length of the recombination region. Each minimal recombination region is trimmed to the first and last k-mer variation shared specifically between the query sequence and the reference sequence(s) within the recombination region. Thus, all the sequences between the two minimal recombination regions are shared between the query sequence and the two sets of reference sequences corresponding to the two minimal recombination regions, except for mutation events; i.e, those sequences are the possible flanking homologous regions with maximal length associated with the recombination.
  
</details>

**Bi-directional k-mer scanning to identify recombination regions**

<details>
  <summary>Click to expand!</summary>
  
</details>
  
  
